File: ingress-service/internal/mq/publisher.go
package mq

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	amqp "github.com/rabbitmq/amqp091-go"

	"yunxiao-ingress-service/internal/model"
)

// Publisher 定义标准事件发布接口，避免业务代码直接绑定 RabbitMQ 实现。
type Publisher interface {
	Publish(ctx context.Context, event model.NormalizedEvent) error
}

// RabbitPublisher 使用 RabbitMQ topic exchange 发布标准事件。
type RabbitPublisher struct {
	conn     *amqp.Connection
	channel  *amqp.Channel
	exchange string
}

var newRabbitPublisher = NewRabbitPublisher

// ConnectRetryOptions 定义 RabbitMQ 启动连接重试策略。
type ConnectRetryOptions struct {
	InitialBackoff time.Duration
	MaxBackoff     time.Duration
	MaxWait        time.Duration
	Sleep          func(time.Duration)
}

// NewRabbitPublisher 创建 RabbitMQ 发布器，并确保目标 exchange 存在。
func NewRabbitPublisher(url string, exchange string) (*RabbitPublisher, error) {
	conn, err := amqp.Dial(url)
	if err != nil {
		return nil, err
	}

	channel, err := conn.Channel()
	if err != nil {
		conn.Close()
		return nil, err
	}

	if err := channel.ExchangeDeclare(exchange, "topic", true, false, false, false, nil); err != nil {
		channel.Close()
		conn.Close()
		return nil, err
	}

	return &RabbitPublisher{conn: conn, channel: channel, exchange: exchange}, nil
}

// NewRabbitPublisherWithRetry 创建 RabbitMQ 发布器，并在启动期连接失败时按指数退避重试。
func NewRabbitPublisherWithRetry(url string, exchange string, options ConnectRetryOptions) (*RabbitPublisher, error) {
	initialBackoff := positiveDuration(options.InitialBackoff, time.Second)
	maxBackoff := positiveDuration(options.MaxBackoff, 10*time.Second)
	maxWait := positiveDuration(options.MaxWait, 60*time.Second)
	sleep := options.Sleep
	if sleep == nil {
		sleep = time.Sleep
	}

	backoff := initialBackoff
	waited := time.Duration(0)
	var lastErr error
	for attempt := 1; ; attempt++ {
		publisher, err := newRabbitPublisher(url, exchange)
		if err == nil {
			if attempt > 1 {
				log.Printf("rabbitmq connected after retry attempts=%d", attempt)
			}
			return publisher, nil
		}
		lastErr = err
		if waited+backoff >= maxWait {
			break
		}
		log.Printf("rabbitmq connect failed attempt=%d backoff=%s error=%q", attempt, backoff, err.Error())
		sleep(backoff)
		waited += backoff
		backoff = nextBackoff(backoff, maxBackoff)
	}
	return nil, fmt.Errorf("connect rabbitmq after %s: %w", maxWait, lastErr)
}

// Publish 使用 event_type 作为 routing key 发布标准事件。
func (p *RabbitPublisher) Publish(ctx context.Context, event model.NormalizedEvent) error {
	body, err := json.Marshal(event)
	if err != nil {
		return err
	}

	return p.channel.PublishWithContext(
		ctx,
		p.exchange,
		event.EventType,
		false,
		false,
		amqp.Publishing{
			ContentType:  "application/json",
			DeliveryMode: amqp.Persistent,
			Timestamp:    time.Now().UTC(),
			MessageId:    event.EventID,
			Headers: amqp.Table{
				"x-trace-id":   event.TraceID,
				"x-event-id":   event.EventID,
				"x-event-type": event.EventType,
				"x-source":     event.Source,
			},
			Body: body,
		},
	)
}

// Close 关闭 RabbitMQ channel 和 connection。
func (p *RabbitPublisher) Close() error {
	if p.channel != nil {
		_ = p.channel.Close()
	}
	if p.conn != nil {
		return p.conn.Close()
	}
	return nil
}

func positiveDuration(value time.Duration, fallback time.Duration) time.Duration {
	if value <= 0 {
		return fallback
	}
	return value
}

func nextBackoff(current time.Duration, maxBackoff time.Duration) time.Duration {
	next := current * 2
	if next > maxBackoff {
		return maxBackoff
	}
	return next
}

File: orchestrator-service/app/publisher.py
import json
import time

import pika


def build_normalized_event_message(row: dict[str, object]) -> dict[str, object]:
    external_refs = row.get("external_refs_json") or {}
    event = {
        "event_id": row["event_id"],
        "source": row["source"],
        "event_type": row["event_type"],
        "trace_id": row["trace_id"],
        "dedup_key": row["dedup_key"],
        "subject": {
            "type": row["subject_type"],
            "id": row["subject_id"],
        },
        "raw_payload_id": f"raw_{row['raw_event_id']}",
    }
    if row.get("project_id"):
        event["project_id"] = row["project_id"]
    if row.get("work_item_id"):
        event["work_item_id"] = row["work_item_id"]
    if external_refs:
        event["external_refs"] = external_refs
    occurred_at = row.get("occurred_at")
    if occurred_at is not None:
        event["occurred_at"] = occurred_at.isoformat()
    return event


def publish_normalized_event(channel, exchange: str, row: dict[str, object]) -> None:
    body = json.dumps(build_normalized_event_message(row), ensure_ascii=False).encode("utf-8")
    channel.basic_publish(
        exchange=exchange,
        routing_key=str(row["event_type"]),
        body=body,
        properties=pika.BasicProperties(
            content_type="application/json",
            delivery_mode=2,
            timestamp=int(time.time()),
            message_id=str(row["event_id"]),
            headers={
                "x-trace-id": str(row["trace_id"]),
                "x-event-id": str(row["event_id"]),
                "x-event-type": str(row["event_type"]),
                "x-source": str(row["source"]),
            },
        ),
    )
